src/mender-update/standalone/standalone.cpp

```diff
@@ -201,618 +222,435 @@ error::Error RemoveStateData(database::KeyValueDatabase &db) {
-		return result;
-	}
+	StateMachine state_machine {ctx};
 
-	auto err = update_module.Cleanup();
+	err = state_machine.SetStopAfterState(ctx.stop_after);
 	if (err != error::NoError) {
-		result.result = Result::FailedAndRollbackFailed;
-		result.err = result.err.FollowedBy(err);
+		return {Result::Failed, err};
 	}
 
-	if (result.result == Result::RolledBack) {
-		err = RemoveStateData(main_context.GetMenderStoreDB());
-	} else {
-		err = CommitBrokenArtifact(main_context, data);
-	}
-	if (err != error::NoError) {
-		result.result = Result::RollbackFailed;
-		result.err = result.err.FollowedBy(err);
-	}
+	state_machine.Run();
 
-	return result;
+	return ctx.result_and_error;
 }
 
-ResultAndError DoInstallStates(
-	context::MenderContext &main_context,
-	StateData &data,
-	artifact::Artifact &artifact,
-	update_module::UpdateModule &update_module) {
-	auto payload = artifact.Next();
-	if (!payload) {
-		return {Result::FailedNothingDone, payload.error()};
-	}
-
-	const auto &default_paths {main_context.GetConfig().paths};
-	events::EventLoop loop;
-	auto script_runner {executor::ScriptRunner(
-		loop,
-		chrono::seconds {main_context.GetConfig().state_script_timeout_seconds},
-		chrono::seconds {main_context.GetConfig().state_script_retry_interval_seconds},
-		chrono::seconds {main_context.GetConfig().state_script_retry_timeout_seconds},
-		default_paths.GetArtScriptsPath(),
-		default_paths.GetRootfsScriptsPath())};
-
-	// ProvidePayloadFileSizes
-	auto with_sizes = update_module.ProvidePayloadFileSizes();
-	if (!with_sizes) {
-		log::Error("Could not query for provide file sizes: " + with_sizes.error().String());
-		return InstallationFailureHandler(main_context, data, update_module);
+ResultAndError Resume(Context &ctx) {
+	auto exp_in_progress = LoadStateData(ctx.main_context.GetMenderStoreDB());
+	if (!exp_in_progress) {
+		return {Result::Failed, exp_in_progress.error()};
 	}
+	auto &in_progress = exp_in_progress.value();
 
-	// Download Enter
-	auto err = script_runner.RunScripts(executor::State::Download, executor::Action::Enter);
-	if (err != error::NoError) {
-		err = err.FollowedBy(
-			script_runner.RunScripts(executor::State::Download, executor::Action::Error));
-		err = err.FollowedBy(update_module.Cleanup());
-		err = err.FollowedBy(RemoveStateData(main_context.GetMenderStoreDB()));
-		return {Result::FailedNothingDone, err};
-	}
-	if (with_sizes.value()) {
-		err = update_module.DownloadWithFileSizes(payload.value());
-	} else {
-		err = update_module.Download(payload.value());
-	}
-	if (err != error::NoError) {
-		err = err.FollowedBy(
-			script_runner.RunScripts(executor::State::Download, executor::Action::Error));
-		err = err.FollowedBy(update_module.Cleanup());
-		err = err.FollowedBy(RemoveStateData(main_context.GetMenderStoreDB()));
-		return {Result::FailedNothingDone, err};
+	if (!in_progress) {
+		return {
+			Result::NoUpdateInProgress,
+			context::MakeError(context::NoUpdateInProgressError, "Cannot resume")};
```

Comment:
Suggestion: elaborate the user message a bit more - "Cannot resume. No update in progress found."

Reply:
It is implicitly elaborated already: The error code will provide a string message similar to what you wrote, once it's printed.